### PR TITLE
GCE firewall rules to be updated when attributes changes.  Fixes #2111.

### LINKED ISSUE #2111: gce_net dont update existing firewall rule
#### Issue type

Bug Report
#### Ansible Version

ansible 1.9.3
#### Environnement

 Ubuntu 15.04
#### Summary

Update firewall rule succeed without doing the job
#### Steps to reproduce

$ ansible localhost -i inventory, -m gce_net -a "name=mynet fwname=testrule allowed=tcp:22"

``` sh
 localhost | success >> {
    "allowed": "tcp:22",
    "changed": true,
    "fwname": "testrule",
    "ipv4_range": "172.17.0.0/24",
    "name": "mynet",
    "src_range": null,
    "src_tags": null,
    "state": "present",
    "target_tags": null
 }
```

$ ansible localhost -i inventory, -m gce_net -a "name=mynet fwname=testrule allowed=tcp:25"

``` sh
 localhost | success >> {
    "allowed": "tcp:25",
    "changed": false,
    "fwname": "testrule",
    "ipv4_range": "172.17.0.0/24",
    "name": "mynet",
    "src_range": null,
    "src_tags": null,
    "state": "present",
    "target_tags": null
 }
```
#### Expected result

The rule is not allowing tcp:22 port but tcp:25 (or the task failled)
#### Actual result

 $ gcloud compute firewall-rules describe testrule

``` sh
allowed:
- IPProtocol: tcp
  ports:
 - '22'
creationTimestamp: '2015-09-18T14:54:07.154-07:00'
description: ''
id: '5595550375581137168'
kind: compute#firewall
name: testrule
network: mynet
sourceRanges:
- 0.0.0.0/0
```

### PR DESCRIPTION
**Pull request by supertom**
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

gce_net
##### ANSIBLE VERSION

```
ansible 2.2.0 (ansible-2111-fix 0e16a5f3ee) last updated 2016/07/06 16:39:09 (GMT +000)
  lib/ansible/modules/core: (ansible-2111-fix d946440bd6) last updated 2016/07/11 16:25:36 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD e0b3e2f790) last updated 2016/07/06 16:23:22 (GMT +000)
  config file = 
  configured module search path = Default w/o overrides

```
##### SUMMARY

Allow GCE firewall rules to be updated when attributes changes. Fixes #2111

Previously, when the attributes of a GCE firewall change, they were ignored.  This PR changes that behavior and now updates them.

Note that the "update" also removes attributes that are not specified.

An overview of the firewall rule behavior is as follows:
1. firewall name in GCP, state=absent in PLAYBOOK: Delete from GCP
2. firewall name in PLAYBOOK, not in GCP: Add to GCP.
3. firewall name in GCP, name not in PLAYBOOK: No change.
4. firewall names exist in both GCP and PLAYBOOK, attributes differ: Update GCP to match attributes from PLAYBOOK.
#### Examples

ansible localhost -i inventory, -m gce_net -a "name=mynet fwname=testrule44 allowed=tcp:24"
(create rule)

```
localhost | SUCCESS => {
    "allowed": "tcp:24", 
    "changed": true, 
    "fwname": "testrule44", 
    "ipv4_range": null, 
    "name": "mynet", 
    "src_range": null, 
    "src_tags": null, 
    "state": "present", 
    "target_tags": null
}
```

ansible localhost -i inventory, -m gce_net -a "name=mynet fwname=testrule44 allowed=tcp:24 src_tags=my-tag"
(add source tag)

```
localhost | SUCCESS => {
    "allowed": "tcp:24", 
    "changed": true, 
    "fwname": "testrule44", 
    "ipv4_range": null, 
    "name": "mynet", 
    "src_range": null, 
    "src_tags": [
        "my-tag"
    ], 
    "state": "present", 
    "target_tags": null
}
```

ansible localhost -i inventory, -m gce_net -a "name=mynet fwname=testrule44 allowed=tcp:22 src_tags=my-tag"
(change port)

```
localhost | SUCCESS => {
    "allowed": "tcp:22", 
    "changed": true, 
    "fwname": "testrule44", 
    "ipv4_range": null, 
    "name": "mynet", 
    "src_range": null, 
    "src_tags": [
        "my-tag"
    ], 
    "state": "present", 
    "target_tags": null
}

```

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @supertom. To the current maintainers, @erjohnso please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by erjohnso** [comment]
shipit

Thanks for the fix @supertom!

/cc @ryansb

**Comment by gregdek** [comment]
Thanks again to @supertom for this PR, and thanks @erjohnso for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/google/gce_net.py [hunk 1]
```diff
@@ -246,12 +247,48 @@ def main():
...
+            # update if attributes changed.  Create if doesn't exist.
             try:
-                gce.ex_create_firewall(fwname, allowed_list, network=name,
+                fw_changed = False
+                fw = gce.ex_get_firewall(fwname)
+
+                # If old and new attributes are different, we update the firewall rule.
+                # This implicitly let's us clear out attributes as well.
+                # allowed_list is required and must not be None for firewall rules.
+                if allowed_list and (allowed_list != fw.allowed):
```
Comment (ryansb) [inline cloud/google/gce_net.py]: Wouldn't this end up spuriously updating if the order of the allowed list changes? I think you need to sort the lists before equality checking.

**Comment by ryansb** [comment]
Other than sorting parameter lists, looks great :+1:

**Comment by supertom** [comment]
@ryansb whatcha think?

**Pull request merged**